nomad/consul_policy.go

```diff
@@ -67,17 +67,43 @@ func (c *consulACLsAPI) isManagementToken(token *api.ACLToken) bool {
 	return false
 }
 
-// namespaceCheck is used to verify the namespace of the object matches the
-// namespace of the ACL token provided.
+// namespaceCheck is used to fail the request if the namespace of the object does
+// not match the namespace of the ACL token provided.
 //
-// exception: iff token is in the default namespace, it may contain policies
+// *exception*: if token is in the default namespace, it may contain policies
 // that extend into other namespaces using namespace_prefix, which must bypass
 // this early check and validate in the service/keystore helpers
+//
+// *exception*: if token is not in a namespace, consul namespaces are not enabled
+// and there is nothing to validate
+//
+// If the namespaces match, whether the token is allowed to perform an operation
+// is checked later.
 func namespaceCheck(namespace string, token *api.ACLToken) error {
-	if token.Namespace != "default" && token.Namespace != namespace {
+
+	switch {
+	case namespace == "" && token.Namespace == "":
+		// ACLs not enabled (if Consul ACLs are enabled, the token response from
+		// consul will always set the namespace (to "default" if otherwise unset))
+		return nil
+
+	case token.Namespace == "default":
+		// ACLs enabled, must defer to per-object checking, since the token could
+		// have namespace_prefix blocks allowing the operation
```

Comment:
I'm not familiar with namespace_prefix. Is it only applicable if token's namespace is default? Why doesn't it apply in the default or `token.Namespace != "default"` cases? Is a special case for handling legacy apps only? Might be useful to add more context.

Reply:
Yep, Consul tokens can supply `namespace` or `namespace_prefix` blocks which give the token permission beyond the "default" namespace. Those blocks are only allowable in tokens in the "default" namespace. 

https://www.consul.io/docs/security/acl/acl-rules#namespace-rules

Expanded on the comment to clarify.